base/system/services/database.c

```diff
@@ -852,6 +852,105 @@ ScmDeleteServiceRecord(PSERVICE lpService)
+
+DWORD ScmDeleteService(PSERVICE lpService)
+{
+    DWORD dwError;
+    DWORD pcbBytesNeeded = 0;
+    DWORD dwServicesReturned = 0;
+    HKEY hServicesKey;
+
+    ASSERT(lpService->RefCount == 0);
+
+    /* Open the Services Reg key */
+    dwError = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                            L"System\\CurrentControlSet\\Services",
+                            0,
+                            KEY_SET_VALUE | KEY_READ,
+                            &hServicesKey);
+    if (dwError != ERROR_SUCCESS)
+    {
+        DPRINT("Failed to open services key\n");
+        return dwError;
+    }
+
+    /* Call the function with NULL, just to get bytes we need */
+    Int_EnumDependentServicesW(hServicesKey,
+                               lpService,
+                               SERVICE_ACTIVE,
+                               NULL,
+                               &pcbBytesNeeded,
+                               &dwServicesReturned);
+
+    /* If pcbBytesNeeded returned a value then there are services running that are dependent on this service */
+    if (pcbBytesNeeded)
+    {
+        DPRINT("Deletion failed due to running dependencies\n");
+        RegCloseKey(hServicesKey);
+        return ERROR_SUCCESS;
+    }
+
+    /* There are no references and no running dependencies,
+       it is now safe to delete the service */
+
+    /* Delete the Service Key */
+    dwError = ScmDeleteRegKey(hServicesKey, lpService->lpServiceName);
+
+    RegCloseKey(hServicesKey);
+
+    if (dwError != ERROR_SUCCESS)
+    {
+        DPRINT("Failed to Delete the Service Registry key\n");
+        return dwError;
+    }
+
+    /* Delete the Service */
+    ScmDeleteServiceRecord(lpService);
+
+    return ERROR_SUCCESS;
+}
+
+/*
+ * This function allows the caller to be sure that the service won't be freed unexpectedly.
+ * In order to be sure that lpService will be valid until the reference is added
+ * the caller needs to hold the database lock.
+ * A running service will keep a reference for the whole time it is not SERVICE_STOPPED.
+ * A service handle will also keep a reference to a service. Keeping a reference is
+ * really needed so that ScmControlService can be called without keeping the database locked.
+ * This means that eventually the correct order of operations to send a control message to 
+ * a service looks like: lock, reference, unlock, send control, lock, dereference, unlock.
+ */
+DWORD
+ScmReferenceService(PSERVICE lpService)
+{
+    return InterlockedIncrement(&lpService->RefCount);
+}
+
+/* This function must be called with the database lock held exclusively as
+   it can end up deleting the service */
+DWORD
+ScmDereferenceService(PSERVICE lpService)
+{
+    ULONG ref;
+
+    ASSERT(lpService->RefCount > 0);
+
+    ref = InterlockedDecrement(&lpService->RefCount);
+#if 0
+    /* Service should not be automatically deleted here. */
+    if (ref == 0 && lpService->bDeleted)
+        ScmDeleteService(lpService);
+#endif
```

Comment:
Does this code make sense if it's disabled anyway? Might it be worth to remove it or it could be kept for reference purposes? :thinking:

Reply:
As it appears, we have only `ScmDeleteServiceRecord`, but no `ScmDeleteService`:
![Снимок экрана от 2024-09-26 19-37-30](https://github.com/user-attachments/assets/d9927384-75cf-436b-8bd0-97778c257c89)
So you probably want to call `ScmDeleteServiceRecord` here, in case you prefer to keep this disabled code for the reference.

Reply:
This code was added by Giannis in PR #1225, but based on my testing it is not quite correct in the use of the 'bDeleted' structure member. When I tried to correct the code and it actually called the ScmDeleteService it removed all of the registry entries for the Spooler service that I was testing from the registry. This was very difficult to recover from and does not happen on W2K3SP2. I wanted to leave it here for discussion but non-functional for the present time. It believe that it should be removed, but I wanted input from others on this. Thanks.

Reply:
Screenshot from my 8+year-old attempt at fixing the same bug as this PR tries to, I have the following code:
![image](https://github.com/user-attachments/assets/da0816ce-5b51-4a72-b1f0-cfb1032638df)

Reply:
Of course my patch in the screenshot was based on an older version of our services.exe code.
Newer code (as you can see on the deleted code in the diff), has:
```c
            /* If this service has been marked for deletion */
            if (lpService->bDeleted &&
                lpService->Status.dwCurrentState == SERVICE_STOPPED)
```
in the condition (the extra SERVICE_STOPPED part); you may need to keep it.

Reply:
Please, let's just leave all this for another PR. Thanks.

Reply:
Uh. You're moving this piece of code around but now we loose the functionality it did? It's your job to investigate the problem, if it was introduced in your pr, and to fix it.

Reply:
OK. Thank you for your explanation. Now fixed.